### PR TITLE
Nav 32635 fix for docker cleanup

### PR DESCRIPTION
**Pull request by NorbertGebicki-TomTom**

### CODE REVIEW

File: cleanup/cleanDockerImages/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+The `cleanDockerImages.properties` file has the following field:
+
+- `dockerRepos`: A list of Docker repositories to clean. If a repo is not in
+  this list, it will not be cleaned.
+
+For example:
+
+``` json
+dockerRepos = ["example-docker-local", "example-docker-local-2"]
```
Comment (martinm82) [inline cleanup/cleanDockerImages/README.md]: Shouldn't this be here updated with the new syntax?

File: cleanup/cleanDockerImages/cleanDockerImages.groovy [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+executions {
+    cleanDockerImages() { params ->
+        def deleted = []
+        def etcdir = ctx.artifactoryHome.etcDir
+        def propsfile = new File(etcdir, "plugins/cleanDockerImages.properties")
+        def config = new ConfigSlurper().parse(propsfile.toURL())
+        def dryRun = params['dryRun'] ? params['dryRun'][0] as boolean : false
+        config.dockerRepos.each{ policySettings ->
+            def repos = policySettings[ 0 ] ? policySettings[ 0 ] as String : ["__none__"]
+            def daysToKeep = policySettings[ 1 ] ? policySettings[ 1 ] as int : 6
```
Comment (martinm82) [inline cleanup/cleanDockerImages/cleanDockerImages.groovy]: Please, do not use magic numbers, and store the number in a descriptive variable instead.

File: cleanup/cleanDockerImages/cleanDockerImages.groovy [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                imagesPathMap[parentId] = []
+            }
+            imagesPathMap[parentId] << [parentRepoPath, childItem.created]
+        }
+        break
+    }
+}
+
+// This method checks if the docker image's manifest has the property
+// "com.jfrog.artifactory.retention.maxDays" for purge
```
Comment (martinm82) [inline cleanup/cleanDockerImages/cleanDockerImages.groovy]: Update the comment to cover the new feature

File: cleanup/cleanDockerImages/CleanDockerImagesTest.groovy [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import groovy.json.JsonSlurper
+import org.apache.http.client.HttpResponseException
+import org.jfrog.artifactory.client.ArtifactoryClientBuilder
+import org.jfrog.artifactory.client.model.repository.settings.impl.DockerRepositorySettingsImpl
+import spock.lang.Specification
+
+class CleanDockerImagesTest extends Specification {
```
Comment (martinm82) [inline cleanup/cleanDockerImages/CleanDockerImagesTest.groovy:7]: Please add a unit test for the new feature

File: cleanup/cleanDockerImages/cleanDockerImages.properties [hunk 5]
```diff
@@ -0,0 +1,4 @@
+dockerRepos = [
+# 		[["repo_name"], number_of_days_to_keep],
+		[["sample-repo"], 30],
```
Comment (martinm82) [inline cleanup/cleanDockerImages/cleanDockerImages.properties]: I am wondering whether we should allow here a null value for number_of_days_to_keep to use only the original way of deleting images (based on properties)

**Review by martinm82 — CHANGES_REQUESTED** [review]

File: cleanup/cleanDockerImages/cleanDockerImages.groovy (hunk 2 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/cleanDockerImages.groovy]: changed to null

File: cleanup/cleanDockerImages/cleanDockerImages.groovy (hunk 3 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/cleanDockerImages.groovy]: done

File: cleanup/cleanDockerImages/CleanDockerImagesTest.groovy (hunk 4 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/CleanDockerImagesTest.groovy:7]: unit tests are not working :D

File: cleanup/cleanDockerImages/cleanDockerImages.properties (hunk 5 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/cleanDockerImages.properties]: done :)

File: cleanup/cleanDockerImages/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+- `dockerRepos`: A list of Docker repositories to clean. If a repo is not in
+  this list, it will not be cleaned.
+
+For example:
+
+``` json
+dockerRepos = [
+// 		["repo_name", number_of_days_to_keep],
+//		if number_of_days_to_keep is empty, LABEL in Dockerfile with maxCount should be set.
```
Comment (martinmigasiewicz-tomtom) [inline cleanup/cleanDockerImages/README.md]: But it is possible to define maxDays and maxCount, right? Maybe both should be mentioned here

File: cleanup/cleanDockerImages/cleanDockerImages.groovy [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+// If not found, number_of_days_to_keep from properties file will be used instead.
+
+def checkDaysPassedForDelete(item, daysToKeep) {
+    def maxDaysProp = "docker.label.com.jfrog.artifactory.retention.maxDays"
+    def oneday = TimeUnit.MILLISECONDS.convert(1, TimeUnit.DAYS)
+    def prop = repositories.getProperty(item.repoPath, maxDaysProp)
+    if (!prop && (daysToKeep == null)) return false
+        else if (((new Date().time - item.created) / oneday) >= daysToKeep) return true  
+    log.debug("PROPERTY $maxDaysProp FOUND = $prop IN MANIFEST FILE")
+    prop = prop.isInteger() ? prop.toInteger() : null
```
Comment (martinmigasiewicz-tomtom) [inline cleanup/cleanDockerImages/cleanDockerImages.groovy:121]: Isn't `toInteger()` returning null if it cannot be converted to an int?

**Review by martinmigasiewicz-tomtom — CHANGES_REQUESTED** [review]

File: cleanup/cleanDockerImages/cleanDockerImages.groovy (hunk 7 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/cleanDockerImages.groovy:121]: I don't know :) This is not mine :)

File: cleanup/cleanDockerImages/README.md (hunk 6 above)
Comment (NorbertGebicki-TomTom) [reply cleanup/cleanDockerImages/README.md]: updated :)

**Review by martinmigasiewicz-tomtom — APPROVED** [review]

**Review by maikelvdh — APPROVED** [review]

**Pull request merged**